core/src/main/kotlin/users/service/UserService.kt

```diff
@@ -195,4 +232,120 @@ class UserServiceImpl(
     override suspend fun update(user: User): User {
         return userRepository.save(user)
     }
+
+    override suspend fun sendVerificationCode(user: User, email: String) {
+        if (user.isEmailVerified == true) throw EmailAlreadyVerifiedException
+        if (!authService.isValidEmail(email)) throw InvalidEmailException
+        if (userRepository.existsByEmailAndIsEmailVerifiedTrueAndActiveTrue(email)) throw DuplicateEmailException
+        val key = "verification-code-${user.id}"
+        val code = (Math.random() * 1000000).toInt().toString().padStart(6, '0')
+        val value = getVerificationValue(email, code, key)
+        val emailSubject = "[SNUTT] 인증번호 [$code] 를 입력해주세요"
+        val emailBody = "<h2>인증번호 안내</h2><br/>" +
+            "안녕하세요. SNUTT입니다. <br/> " +
+            "<b>아래의 인증번호 6자리를 진행 중인 화면에 입력하여 3분내에 인증을 완료해주세요.</b><br/><br/>" +
+            "<h3>인증번호</h3><h3>$code</h3><br/><br/>" +
+            "인증번호는 이메일 발송 시점으로부터 3분 동안 유효합니다."
+        redisTemplate.opsForValue().set(key, mapper.writeValueAsString(value), Duration.ofMinutes(3)).subscribe()
+        mailClient.sendMail(email, emailSubject, emailBody)
+    }
+
+    override suspend fun verifyEmail(user: User, code: String) {
+        val key = "verification-code-${user.id}"
+        val value = mapper.treeToValue(mapper.reader().readTree(redisTemplate.opsForValue().getAndAwait(key)), RedisVerificationValue::class.java)
+        if (value == null || value.code != code) throw InvalidVerificationCodeException
+        user.email = value.email
+        user.isEmailVerified = true
+        userRepository.save(user)
+        redisTemplate.delete(key).subscribe()
+    }
+
+    override suspend fun resetEmailVerification(user: User) {
+        user.isEmailVerified = false
+        userRepository.save(user)
+    }
+
+    override suspend fun attachLocal(user: User, localLoginRequest: LocalLoginRequest) {
+        if (user.credential.localId != null) throw AlreadyLocalAccountException
+        val localId = localLoginRequest.id
+        val password = localLoginRequest.password
+        if (!authService.isValidLocalId(localId)) throw InvalidLocalIdException
+        if (!authService.isValidPassword(password)) throw InvalidPasswordException
+        if (userRepository.existsByCredentialLocalIdAndActiveTrue(localId)) throw DuplicateLocalIdException
+        val localCredential = authService.buildLocalCredential(localId, password)
+        user.credential.localId = localCredential.localId
+        user.credential.localPw = localCredential.localPw
+        user.credentialHash = authService.generateCredentialHash(user.credential)
```

Comment:
kotlin apply 사용해서 데이터 수정이 어디서 일어나는지 한눈에 알아볼 수 있으면 좋을듯요
```suggestion
        user.apply {
            credential.localId = localCredential.localId
            credential.localPw = localCredential.localPw
            credentialHash = authService.generateCredentialHash(user.credential)
        }
```